contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,656 @@
+              
+               Milestone 3: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 4: Expose volumeMode to users via kubectl
+               
+               Milestone 5: PV controller binding changes for block devices
+               	       
+	       Milestone 6: Container Runtime changes
+	       
+	       Milestone 7: Initial Plugin changes (GCE, AWS & GlusterFS)
+	       
+	       Milestone 8: Disabling of provisioning where volumeMode == Block is not supported
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume update
+		
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice(podUID types.UID) error
+       SetUpDeviceAt(dir string, podUID types.UID) error
+       GetDevicePath() (string, error)
+       GetVolumeDeviceMapPath(spec *Spec) (string, error)
+ }
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+       GetVolumeDeviceUnmapPath(spec *Spec) (string, error)
+ }
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/
+
+- Volume mount path
+/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link to map raw block device.
+Kubelet creates a new symbolic link under the new global map path when volume is attached to a Pod. Number of symbolic links
+are equal to the number of Pods which uses the same volume. Kubelet needs to manage both creation and deletion of symbolic links
+under the global map path. The name of the symbolic link is same as pod uuid.
+ 
+```
+Global map path for "Block" volumeMode volume
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/pod-uuid1
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/pod-uuid2
+```
```

Comment:
If it is global path, why does it has pod-uuid in it?

Reply:
This is needed to refcnt how many pods are using a block volume.  For filesystem volumes, it's done by counting the mounts, but there is no equivalent method for block devices.

Reply:
so the string "pod-uuid" is not pod uuid right? it is is some-string+"count_of_pods_using_it" ?

Reply:
No, it's a pod's individual uuid.
May be {pod uuid} is better to avoid confusion?

Reply:
Yeah that will def. make it clearer. Also what is {volumePluginDependentPath} path btw?

Reply:
Ex.

FC plugin:   volumeDevices/fc/{pod uuid}
Iscsiplugin:  volumeDevices/iscsi/iface-default/{pod uuid}

Reply:
We already have same path(directory) for file system volume now.